rfcs/0058-secrets-for-services.md

```diff
@@ -0,0 +1,273 @@
+ways across the ecosystem. There have also been ideas like adjustments to the
+Nix Store (like issue https://github.com/NixOS/nixpkgs/issues/8), which
+would allow for non-world-readable files, but this issue has made no progress
+in several years.
+
+With the introduction of Systemd's `DynamicUser`, the more traditional
+approaches of manually managing permissions of some out-of-store files could
+become cumbersome or slow down the adoption of DynamicUser and other sandboxing
+features throughout the nixpkgs modules.
+
+The approach outlined in this document aims to solve only a part of the secrets
+management problem, namely: How to make secrets that are already accessible on the
+system (be it through a secrets folder only readable by root, or a system like
+vault or nixops) available to non-interactive services in a safe way.
+
+It assumes that shipping secrets is already solved sufficiently by krops, nixops,
+git-crypt, simple rsync etc, and if not, that this can be addressed as a separate
+concern without needing to change the approach proposed here. Further, it is outside
+of the scope of this proposal to ensure other properties of the secret store, such as
+encryption at rest.
+
+The main idea here is to allow for flexibility in the way secrets are delivered to the
+system, while at the same time providing a consistent and unobtrusive mechanism that can
+be applied widely across service modules without requiring large code-changes while allowing
+for a gradual transition of nixos services.
+
+# Detailed design
+[design]: #detailed-design
+
+To summarize, necessary preconditions:
+
+* Delivery of secrets to target systems is a solved problem
+* It's sufficiently secure to store the secrets or access tokens in a location
+  only accessible by root on the system
+* The secrets store locations is secure at rest, such as full-disk-encryption.
+* Interactive unlocking scenarios are should be treated separately
+* Linux namespaces are sufficiently secure
+* The service can be run using `PrivateTmp`
+
+Design goals:
+* A set of secrets are made available to a set of services only for the duration of their execution
+* Retrieved secrets are only accessible to the service processes and root
+* Retrieved secrets are reliably cleaned up when the services stop, crash,
+  receive sigkill or the system is restarted
+
+Core concepts and terminology:
+
+* *Secrets store*: a secure file-system based location, in this document
+  `/etc/secrets`, only accessible to root
+* A *fetcher* function: a function whose task it is to resolve the secret
+  identifier, retrieve the secret and place it in the service process' private
+  namespace within /tmp name
+* Simple helper functions to *enrich* expressions defining systemd services
+  with secrets
+* "Side-car" service: A privileged systemd service running the fetcher
+  function to retrieve the fetcher function, and initially create the service
+  namespace
```

Comment:
"Running the fetcher function to retrieve the fetcher function"? Should one of the two mentions of "fetcher function" mean something else?

Reply:
I'll resolve this for now, since I accepted the simplified suggestion below. If it turns out to be confusing, I'll rephrase. It rather obvious from the example code though.